kotsadm/pkg/gitops/key.go

```diff
@@ -0,0 +1,43 @@
+package gitops
+
+import (
+	"crypto/rsa"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+var (
+	// Generating the gitops key is very slow (~10s).
+	// Pre-seed it to save time in the UI.
+	preseedPrivateKey   *rsa.PrivateKey
+	preseedPrivateKeyMu sync.Mutex
+)
+
+var r *rand.Rand
+
+func init() {
+	r = rand.New(rand.NewSource(time.Now().UTC().UnixNano()))
+
+	go generatePreseedPrivateKey()
+}
+
+func getPrivateKey() (*rsa.PrivateKey, error) {
+	preseedPrivateKeyMu.Lock()
+	key := preseedPrivateKey
+	preseedPrivateKey = nil // one time use
+	preseedPrivateKeyMu.Unlock()
+
+	go generatePreseedPrivateKey()
+
+	if key != nil {
+		return key, nil
+	}
+	return rsa.GenerateKey(r, 4096)
+}
+
+func generatePreseedPrivateKey() {
+	preseedPrivateKeyMu.Lock()
+	defer preseedPrivateKeyMu.Unlock()
+	preseedPrivateKey, _ = rsa.GenerateKey(r, 4096)
```

Comment:
why not handle/log the error here?

Reply:
i could panic? this is in a goroutine in init() and it would be hard to surface this to the user

Reply:
also, the failure mode here is 'generate another cert later, and return the error there if present'